guides/technical-guides/test-driven-development.md

```diff
@@ -0,0 +1,105 @@
+# Test Driven Development (TDD)
+
+It is expected that where feasible and appropriate, all projects will have
+relevant tests and test automation to accompany the functional elements of the
+project.
+
+One great way to ensure this practice is adhered to is through the medium of
+**Test Driven Development**, commonly referred to as **TDD**.
+
+## What is TDD?
+
+TDD is a method of development whereby (in an _ideal_ world) unit tests are
+written before the application code of the project.
+
+The ideal flow of TDD is as follows:
+
+1. Write a failing test
+2. Write functional code to make the test pass
+3. Yay, we have passing tests!
+
+## Why TDD?
+
+The advantage of writing code in a TDD way is that one can set out exactly what
+should make a test pass and what shouldn't.
+
+Writing the application code after the test means that the bare minimum code is
+written to fulfil the requirement for a particular test. Additionally, the
+theory goes that writing the test beforehand means the test shouldn't be forced
+to pass by writing complex test logic.
+
+This helps to reduce the complexity of individual methods and can reduce scope
+creep for particular requirements.
+
+TDD also helps to reduce bugs in the code as individual methods are tested only
+for what they are written to achieve. This is great because the testing is
+low-level, meaning that differing conditions at runtime _probably_ won't hide
+other issues.
+
+## Example (in Python 🐍)
+
+_Assume the relevant `unittest` packages are imported and setup._
+
+1. Write a test that should pass, but fails because the method isn't implemented
+   yet.
+
+   `test_calculator.py`
+
+   ```python
+   def test_calculator_add(self):
+       self.assertEquals(Calculator.add(2, 2), 4)
+   ```
+
+2. Write the method to make the test pass.
+
+   `calculator.py`
+
+   ```python
+   def add(self, a, b):
+       return a + b;
+   ```
+
+You can see in the above that we start by writing a test which asserts that
+`2 + 2 = 4`. We then write our application method that will return `4` when `2`
+is passed to both parameter `a` and parameter `b`.
+
+## Useful Tools
```

Comment:
Which bit? I'm not sure GitHub is showing me the right bit?

Reply:
The "Useful Tools" section.

They're _testing frameworks_ rather than something that's TDD specific

Reply:
I get what you mean, figured it would be useful to add here as examples of things you can make use of for implementing TDD practically in a project. It could be moved to testing but I wanted this page to have some practical information, too